conans/client/generators/cmake_find_package.py

```diff
@@ -43,52 +48,297 @@ class CMakeFindPackageGenerator(Generator):
-
-        # The find_libraries_block, all variables for the package, and creation of targets
-        public_deps_names = [self.deps_build_info[dep].get_name("cmake_find_package") for dep in
-                             dep_cpp_info.public_deps]
-        deps_names = ";".join(["{n}::{n}".format(n=n) for n in public_deps_names])
-
-        deps = DepsCppCmake(dep_cpp_info)
-        find_libraries_block = target_template.format(name=name, deps=deps, build_type_suffix="",
-                                                      deps_names=deps_names)
-
-        # The find_transitive_dependencies block
-        find_dependencies_block = ""
-        if dep_cpp_info.public_deps:
-            # Here we are generating FindXXX, so find_modules=True
-            f = find_transitive_dependencies(public_deps_names, find_modules=True)
-            # proper indentation
-            find_dependencies_block = ''.join("        " + line if line.strip() else line
-                                              for line in f.splitlines(True))
-
-        tmp = self.find_template.format(name=name, version=dep_cpp_info.version,
-                                        find_libraries_block=find_libraries_block,
-                                        find_dependencies_block=find_dependencies_block,
-                                        macros_and_functions=macros_and_functions)
-        return tmp
+    def _get_components(self, pkg_name, pkg_findname, cpp_info, reverse=True):
+        find_package_components = []
+        for comp_name, comp in self.sorted_components(cpp_info).items():
+            comp_findname = self._get_name(cpp_info.components[comp_name])
+            deps_cpp_cmake = DepsCppCmake(comp)
+            deps_cpp_cmake.public_deps = self._get_component_requires(pkg_name, pkg_findname, comp)
+            find_package_components.append((comp_findname, deps_cpp_cmake))
+        if reverse:
+            find_package_components.reverse()  # From the less dependent to most one
+        return find_package_components
+
+    def _get_component_requires(self, pkg_name, pkg_findname, comp):
+        comp_requires_findnames = []
+        for require in comp.requires:
+            if COMPONENT_SCOPE in require:
+                comp_require_pkg_name = require[:require.find(COMPONENT_SCOPE)]
+                if comp_require_pkg_name not in self.deps_build_info.deps:
+                    raise ConanException("Component '%s' not found: '%s' is not a package "
+                                         "requirement" % (require, comp_require_pkg_name))
+                comp_require_pkg = self.deps_build_info[comp_require_pkg_name]
+                comp_require_pkg_findname = self._get_name(comp_require_pkg)
+                comp_require_comp_name = require[require.find(COMPONENT_SCOPE)+len(COMPONENT_SCOPE):]
+                if comp_require_comp_name in self.deps_build_info.deps:
+                    comp_require_comp_findname = comp_require_pkg_findname
```

Comment:
Is this possible? Will this lead to unexpected behavior if a component has the same name as other required package?

Reply:
This is covering the case of a component depending on the global package. A component may have the same name of a package, but the syntax will be different, so there is no unexpected behavior

Reply:
What I see here is:

```python
class Recipe:
    requires = "mypkg/version", "zlib/1.2.11"

    def package_info(self):
        self.cpp_info.components["cmp"].requires = ["mypkg::zlib"]
        ...
```

From the line `comp_require_pkg_name, comp_require_comp_name = require.split(COMPONENT_SCOPE)`, I would get `comp_require_comp_name` equal to `zlib`, and it'd enter the `if` branch because `zlib` is inside `deps_build_info.deps`... and it doesn't make sense for me.

Reply:
ok, now I get it. I am making a fix for this case